### DISCUSSION TITLE
Adopt clx-truetype?

### DISCUSSION THREAD

**Discussion by PuercoPop** [stumpwm/stumpwm#1109] [General]
The upstream repo of clx-truetype was deleted by its author a couple of months ago for unknown reasons. I've found a fork that _preserves_ the git history at https://github.com/l04m33/clx-truetype

@dmb2 Would you be on board with adopting the repo under the StumpWM organization? afaict we are its main users so it would be convenient to be able to patch any bugs we may find.

**Comment by dmb2** [comment]
I'm not opposed, Mikhail Filonenko is the author of clx-truetype, and also did some foundational work getting xkb off the ground in a pure common lisp fashion.  He was happy to have me take of the xkb portion of his work, so I'm guessing he would be amenable to our adopting clx-truetype.

My major issues with clx-truetype were its performance (Mikhail was clear that performance wasn't a top priority) as well as the fact that it only works with TrueType (TTF) fonts.  I wrote a similar renderer (in fact I borrowed a significant amount of the API Mikhail developed in mine) using FreeType2 through CFFI in order to add support for OpenType fonts (OTF). 

Font rendering is apparently expensive, and my solution seemed to be inherit the same performance problems as clx-truetype with some additional instabilities thrown in.   

I would like to solve the performance problem once and for all, and I would really like to have the ability to use OpenType format fonts. I could move both to the stumpwm organization and let them both live in contrib modules.  

A better, and more ideal solution would be to have one de-facto standard with enough performance to make it the default renderer and ship it with stumpwm proper.  It would go far toward making stumpwm look a little bit more modern, and not like something out of the early 90s.

**Comment by PuercoPop** [comment]
Unless I'm misunderstanding you we are talking about two separate issues, although they are related.

The more pressing issue is that Mikhail Filonenko has deleted their GitHub repository of clx-truetype. Currently there is no official upstream source and clx-truetype is a dependency for the ttf-fonts contrib module (which I use). I'm proposing that we use the stumpwm organization to host clx-truetype for our users to solve this impasse.

If https://github.com/dmb2/clx-ft2-renderer is API compatible with clx-truetype then we could change the ttf-fonts to use that instead. Is that the case? If so would you prefer that? I would.

<hr />

The other issue is how to improve the font support in StumpWM. In that area I'm not sure what would be the best way forward. On that I don't know what would be the best way forward. afaiu xrender should use the GPU to render the fonts, other than that I'm unsure how we could improve the font rendering performance. 

@lokedhs has done some work improving Font rendering in McCLIM, maybe they can weight in on the matter?

**Comment by jorams** [comment]
Since the original repository is entirely gone, getting a copy of it put up is a good idea. Has anyone contacted @filonenko-mikhail about any of this?

**Comment by PuercoPop** [comment]
> Has anyone contacted @filonenko-mikhail about any of this?

I have not, I assumed they had their reasons for deleting the repo as it was the only Lisp one they deleted.

**Comment by lokedhs** [comment]
On Sun, 22 Dec 2019, 02:11 Javier Olaechea, <notifications@github.com>
wrote:

> Has anyone contacted @filonenko-mikhail
> <https://github.com/filonenko-mikhail> about any of this?
>
> I have not, I assumed they had their reasons for deleting the repo as it
> was the only Lisp one they deleted.
>

I have worked on McCLIM which uses cl-ttf (if I remember the name
correctly). It's a pure Lisp implementation of a TTF raseriser.

Actual drawing is done using Xrender, which is quite efficient and easy to
understand. The amount of code needed to get it to work in Stumpwm should
be pretty small.

I don't have time to work on it myself right now, but I'll be happy to
advice anyone who wants to do it.

>

**Comment by jorams** [comment]
It looks like the McCLIM font rendering code is here: https://github.com/McCLIM/McCLIM/tree/016a339946d2b9ba42028ffc5db3e5f1db7b74e2/Extensions/fonts.

It uses `zpb-ttf` and `cl-vectors`, also used by `clx-truetype`.

**Comment by lokedhs** [comment]
On Sun, 22 Dec 2019, 21:37 Joram Schrijver, <notifications@github.com>
wrote:

> It looks like the McCLIM font rendering code is here:
> https://github.com/McCLIM/McCLIM/tree/016a339946d2b9ba42028ffc5db3e5f1db7b74e2/Extensions/fonts
> .
>
> It uses zpb-ttf and cl-vectors, also used by clx-truetype.
>
Yes. Specifically you want to look at xrender-fonts.lisp.

The same directory contains the Freetype backend which is written by me. It
has better quality, but is more complex and also relies on CFFI (and
cffi-grovel), which is not really appropriate for Stumpwm.

>

**Comment by dmb2** [comment]
@lokedhs Thanks for this contribution! It looks like you've got the same dependencies as clx-truetype.  It shouldn't be too much work to mash the two interfaces together.  Again, I hate the fact that the solution depends on TTF fonts and doesn't support FreeType/OpenType fonts, but if someone can deliver an api compatible solution, I'm all game for hosting and maintaining it in the StumpWM organization.

**Comment by lokedhs** [comment]
On Mon, 23 Dec 2019, 04:38 David, <notifications@github.com> wrote:

> @lokedhs <https://github.com/lokedhs> Thanks for this contribution! It
> looks like you've got the same dependencies as clx-truetype. It shouldn't
> be too much work to mash the two interfaces together. Again, I hate the
> fact that the solution depends on TTF fonts and doesn't support
> FreeType/OpenType fonts, but if someone can deliver an api compatible
> solution, I'm all game for hosting and maintaining it in the StumpWM
> organization.
>

You can get Freetype working. McCLIM had two font rending backends that you
can choose from. The first one is the TTF renderer that we've just tallest
about.

The second is the Freetype renderer which I implemented because of the
limitations that you were implying. As part of that work I implemented CFFI
wrappers around Fontconfig and Harfbuzz. The code for the Freetype backend
is in the same directory under the name freetype.lisp if I remember
correctly. The CFFI wrappers can be found in the Extensions directory.

Using these would be a perfectly reasonable alternative, but the CFFI
wrappers need work. Specifically, they need to changed so that they don't
depend on cffi-grovel anymore. This is because cffi-grovel requires the
user to have a C compiler installed as well as all the development headers
for Freetype, Fontconfig and Harfbuzz. That's a bit of a tall order for a
regular user of Stumpwm.

>

**Comment by JMC-design** [comment]
If you guys decide to go stick with clx-truetype, there's a bug in text-line-bounding-box where xmin is hardcoded to 0. I found it when packing up the ripped out guts which you might be interested in as well.  
https://github.com/JMC-design/truetype-clx , that produces just alphamasks which shouldn't be hard to feed to an xrender glyph set. 

n.b. clx is hardcoded for alpha mask glyphsets right now, no other format besides a8 allowed.  If anybody is going to work on a xrender glyph set renderer please tell me so I can skip that and hopefully just plug in what you guys write.

**Comment by dmb2** [comment]
Lets adopt clx-truetype for now so that the module keeps working, its the more popular one anyway.  Long-term, I would like to explore using McCLIM for gui elements and piggy back of their text rendering.  This would help centralize internationalized input development efforts as well.

**Comment by filonenko-mikhail** [comment]
It's seems that I removed clx-truetype in some "gabrage collection" process.

I think using https://github.com/JMC-design/truetype-clx is appropriate.

On Wed, Jan 15, 2020 at 4:51 AM David <notifications@github.com> wrote:

> Lets adopt clx-truetype for now so that the module keeps working, its the
> more popular one anyway. Long-term, I would like to explore using McCLIM
> for gui elements and piggy back of their text rendering. This would help
> centralize internationalized input development efforts as well.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stumpwm/stumpwm/issues/659?email_source=notifications&email_token=AAC2O556SDTVP4LPWP723Q3Q5ZT3ZA5CNFSM4JOIHAWKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEI6YTPA#issuecomment-574458300>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC2O55DBS62PAFXVFBQTFLQ5ZT3ZANCNFSM4JOIHAWA>
> .
>


-- 
With best regards, Michael Filonenko

**Comment by JMC-design** [comment]
Perhaps my clever naming wasn't.  TRUETYPE-CLX is CLX-TRUETYPE  without the CLX. So a glyphset creator is needed and then a simple function to render the glyphs. 
The current issue I have with creating a glyphset is I haven't found  a clean way of getting the dimensions for spaces. So I put it on the backburner while I finish up console stuff.

**Comment by podiki** [comment]
I think quicklisp just removed clx-truetype, so this has become more urgent. In case this bites anyone else, here are instructions for rolling back quicklisp (the last dist from 2019 is fine): http://blog.quicklisp.org/2011/08/going-back-in-dist-time.html

(Note that I had to reload/compile things like stumpwm, swank, etc. as things didn't want to work at first. All seems good now, but probably easier to just grab the package and put it in local-projects.)

**Comment by PuercoPop** [comment]
> Lets adopt clx-truetype for now so that the module keeps working, its the more popular one anyway.

:+1:, I can't create a repository under the StumpWM organization. @dmb2 Can you create one or fork  https://github.com/l04m33/clx-truetype. After that I'll submit the issue to quicklisp

**Comment by m-g-r** [comment]
UPDATE: I've just made this a separate issue as I should have right away: https://github.com/stumpwm/stumpwm/issues/779

> Long-term, I would like to explore using McCLIM for gui elements and piggy back of their text rendering.

~~@dmb2 hey, that sounds interesting, I was quite active in McCLIM 12 to 15 years ago, contributed some extensions and also worked on Beirc and Climacs, and recently looked a bit into it again. Actually, I was confused by the command system of StumpWM quite a bit because I am more familiar with CLIM and Common Lisp and only wrote a few things for Emacs (e.g. https://github.com/xquery-mode/oook-selector). I'd have expected that the command system of StumpWM would be more like CLIM but it has more an Emacs feeling to it and there are things that are necessary maybe in Emacs Lisp but not in Common Lisp in it. Well, the influence of Emacs is, of course, apparent when you look a the mode-line idea for example.~~

~~So, what do you think about adopting the commands and presentations of CLIM?~~

~~PS: Of course, this is a different issue. If you buy into this idea we could make this a new issue.~~

**Comment by jdormit** [comment]
@PuercoPop @dmb2 Thanks for your work here so far! Any progress on uploading the forked clx-truetype to Quicklisp? I hadn't updated my Quicklisp distribution in far too long and was unpleasantly surprised when the ttf-fonts contrib module broke...

**Comment by jcguu95** [comment]
I can help submit `clx-truetype` to quicklisp if that's legal. Does anyone know how to be sure of that?

**Comment by goose121** [comment]
I'd be willing to submit my patched version from #861 to quicklisp; I don't know that there's any surefire way to know it's legal, seeing as the library doesn't include a license, but... given that it's been forked before it's probably fine

**Comment by PhilHudson** [comment]
Can someone please ELI5 what to do with the code from https://github.com/l04m33/clx-truetype to make a broken stumpwm work? (Manjaro, SBCL 2.1.7, QuickLisp StumpWM 20210807)

Edit: I've git clone'd it. 'make' fails at `(require :sb-texinfo)`

**Comment by jorams** [comment]
@PhilHudson 

1. Clone it into `~/quicklisp/local-projects/`
2. (Not entirely sure if this is necessary, but it doesn't hurt) `sbcl --eval '(ql:register-local-projects)' --quit`

**Comment by PhilHudson** [comment]
@jorams Thank you!

**Comment by goose121** [comment]
I'd suggest using my fork at https://github.com/goose121/clx-truetype,
otherwise you'll run into a memory leak that will eventually crash stumpwm

**Comment by PhilHudson** [comment]
@goose121 Thank you too.

**Comment by dmb2** [comment]
I just forked @goos121's version.  If we want to submit this to quicklisp for the community I'm fine with it!

**Reply by KonstantinDjairo** [reply]
that'd be nice, stumpwm with ttf support would be really cool

**Reply by goose121** [reply]
Please do! That would be great c: I'd also be willing to maintain it, and submit my version to quicklisp; whatever you'd prefer.

**Comment by MIvanchev** [comment]
Good day. Maybe we can settle for 1 fork and begin versioning that. I want to package StumpWM's contrib module for Void Linux but it's not possible without clarity what fork is considered the good one and what version it currently has. Maybe we can settle for @goose121 and make that 1.0.

**Reply by goose121** [reply]
As far as I'm aware, my fork has all of the changes I know of, plus a vital patch to fix memory leaks that will eventually crash the window manager if it displays enough different strings. Additionally, if anyone knows of other fixes that are not yet included in my fork, I am willing to pull them in.

In my opinion, the best option is to use my fork (or dmb2's version of it), and as for versioning, 1.0 sounds good, I don't think it ever had a proper version number. Would you like me to create a release on my repo for this purpose?

**Reply by MIvanchev** [reply]
Yes, from what I've seen @goose121 your fork is the most up-to-date one. I've also been using it for some time and it's working great!

>Would you like me to create a release on my repo for this purpose?

Let me contribute some text for your README.md explaining how we came to this situation and then you could tag and release 1.0 yes.

**Reply by MIvanchev** [reply]
@goose121 I've opened a PR and also noticed the Issues are visible on any of the forks so maybe you can enable them for your repo.

**Reply by goose121** [reply]
Thanks! I've merged your PR and created a release for 1.0.0

**Reply by MIvanchev** [reply]
I hope we are on a good path now :)

**Reply by MIvanchev** [reply]
@goose121 I wasn't clear in my previous message but I think you should also enable the Issues section so we can accept bugs, suggestions etc.

**Reply by goose121** [reply]
oops! yeah that's a good idea, I'll go figure out how to do that :D definitely better with Git than GitHub haha

**Comment by MIvanchev** [comment]
I'm chimning in to say that I've written a new module called `sdl-fonts` that uses SDL2 (and soon maybe 3) for font rendering, the MR is here https://github.com/stumpwm/stumpwm-contrib/pull/306. It doesn't use anything but CFFI and SDL so minimal Lisp dependencies.
